File: Implementation/p4.py
nx  = 0
ny = 0
tmp = 0
cnt = 1
mx = [-1,0,1,0]
my = [0,-1,0,1]
def left():
    global direction
    if direction == 3:
        direction = 0
    else:
        direction += 1
n, m = map(int, input().split())
x, y, direction = map(int, input().split())
a = [[0] * m for i in range(n)]
b = [[0] * m for i in range(n)]
for i in range(n):
    a[i] = list(map(int, input().split()))
b[x][y] = 1
while(True):
    left()
    nx = x + mx[direction]
    ny = y + my[direction]
    if a[nx][ny] == 0 and b[nx][ny] == 0:
        tmp = 0
        cnt += 1
        x = nx
        y = ny
        b[x][y] = 1
        continue
    else:
        tmp += 1
    if tmp == 4:
        nx  = x - mx[direction]
        ny = y - my[direction]
        if a[nx][ny] == 0:
            x = nx
            y = ny
            tmp = 0
        else:
            break
print(cnt)

File: Implementation/p4.cpp
#include <bits/stdc++.h>
using namespace std;
int d;
int a[100][100];
int b[100][100];
int mx[4] = {-1,0,1,0};
int my[4] = {0,-1,0,1};
int n, m;
int x, y;
int nx, ny;
int cnt = 1;
int tmp;
void left(){
    if(d ==3 ){
        d = 0;
    }
    else{
        d += 1;
    }
}
int main(){
    cin >> n >> m;
    cin >> x >> y >> d;
    b[x][y] = 1;
    for(int i = 0; i < n; i++){
        for(int j = 0; j < m; j++){
            cin >> a[i][j];
        }
    }
    while(true){
        left();
        nx = x + mx[d];
        ny = y + my[d];
        if(nx < n && ny < m){
            if(a[nx][ny] == 0 && b[nx][ny] == 0){
                tmp = 0;
                cnt += 1;
                x = nx;
                y = ny;
                b[x][y] = 1;
            }
            else{
            tmp++;
        }
        }
        if(tmp == 4){
            nx = x - mx[d];
            ny = y - my[d];
            if(nx < n && ny < m){
                if(a[nx][ny] == 0){
                x = nx;
                y = ny;
                }
                else{
                break;
                }

            }

        }
    }
    cout << cnt;
}